internal/config/rule.go

```diff
@@ -1,195 +1,127 @@
-				if p := r.TimePeriods[pendingRule.TimePeriodID.Int64]; p == nil {
-					ruleLogger.Warnw("ignoring rule with unknown timeperiod_id")
-					continue
-				} else {
-					pendingRule.TimePeriod = p
-				}
+			elementRule.Escalations[newElement.ID] = newElement
+			return nil
+		},
+		func(curElement, update *rule.Escalation) error {
+			if curElement.RuleID != update.RuleID {
+				return errRemoveAndAddInstead
 			}
 
-			for _, escalation := range pendingRule.Escalations {
-				for i, recipient := range escalation.Recipients {
-					recipientLogger := r.logger.With(
-						zap.Int64("id", recipient.ID),
-						zap.Int64("escalation_id", recipient.EscalationID),
-						zap.Int64("channel_id", recipient.ChannelID.Int64),
-						zap.Inline(recipient.Key))
-
-					if recipient.ContactID.Valid {
-						id := recipient.ContactID.Int64
-						if c := r.Contacts[id]; c != nil {
-							recipient.Recipient = c
-						} else {
-							recipientLogger.Warnw("ignoring unknown escalation recipient")
-							escalation.Recipients[i] = nil
-						}
-					} else if recipient.GroupID.Valid {
-						id := recipient.GroupID.Int64
-						if g := r.Groups[id]; g != nil {
-							recipient.Recipient = g
-						} else {
-							recipientLogger.Warnw("ignoring unknown escalation recipient")
-							escalation.Recipients[i] = nil
-						}
-					} else if recipient.ScheduleID.Valid {
-						id := recipient.ScheduleID.Int64
-						if s := r.Schedules[id]; s != nil {
-							recipient.Recipient = s
-						} else {
-							recipientLogger.Warnw("ignoring unknown escalation recipient")
-							escalation.Recipients[i] = nil
-						}
-					} else {
-						recipientLogger.Warnw("ignoring unknown escalation recipient")
-						escalation.Recipients[i] = nil
-					}
-				}
+			curElement.ChangedAt = update.ChangedAt
+			curElement.NameRaw = update.NameRaw
+			// Condition{,Expr} are being initialized by config.IncrementalConfigurableInitAndValidatable.
+			curElement.Condition = update.Condition
+			curElement.ConditionExpr = update.ConditionExpr
+			// TODO: synchronize Fallback{ForID,s} when implemented
+
+			return nil
+		},
+		func(delElement *rule.Escalation) error {
+			elementRule, ok := r.Rules[delElement.RuleID]
+			if !ok {
+				return nil
+			}
 
-				escalation.Recipients = slices.DeleteFunc(escalation.Recipients, func(r *rule.EscalationRecipient) bool {
-					return r == nil
-				})
+			delete(elementRule.Escalations, delElement.ID)
+			return nil
+		})
+
+	incrementalApplyPending(
+		r,
+		&r.ruleEscalationRecipients, &r.configChange.ruleEscalationRecipients,
+		func(newElement *rule.EscalationRecipient) error {
+			ok := false
+			if newElement.ContactID.Valid {
+				newElement.Recipient, ok = r.Contacts[newElement.ContactID.Int64]
+			} else if newElement.GroupID.Valid {
+				newElement.Recipient, ok = r.Groups[newElement.GroupID.Int64]
+			} else if newElement.ScheduleID.Valid {
+				newElement.Recipient, ok = r.Schedules[newElement.ScheduleID.Int64]
+			}
```

Comment:
`EscalationRecipient` embeds `recipient.Key`, so you should be able to use the following function here right away:
https://github.com/Icinga/icinga-notifications/blob/32bf26e3217a64697faab1ce9f8497a00e0a9517/internal/config/runtime.go#L112-L132